File: src/types/Continent.ts
import { Entity, Object, Required, String } from "tsondb/schema/dsl"
import { NestedTranslationMap } from "./Locale.js"

export const Continent = Entity(import.meta.url, {
  name: "Continent",
  namePlural: "Continents",
  comment:
    "Continents are mostly referenced to in languages and scripts that occur on a specific continent.",
  type: () =>
    Object({
      translations: NestedTranslationMap(
        Required,
        "Continent",
        Object({
          name: Required({
            comment: "The continent name.",
            type: String({ minLength: 1 }),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

File: Sources/OptolithDatabaseSchema/Types/Continent.swift
import FileDB

/// Continents are mostly referenced to in languages and scripts that occur on a specific continent.
@Model
public struct Continent {
    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation { // ContinentTranslation

        /// The continent name.
        @MinLength(1)
        let name: String
    }
}
